src/http_server.c

```diff
@@ -0,0 +1,234 @@
+		return -1;
+	}
+
+	// Safety Null-terminate
+	request[sizeof(request) - 1] = 0;
+	request[read_result] = 0;
+
+	static const char *request_head = "GET /?id=";
+
+	if (strncmp(request, request_head, strlen(request_head)) != 0) {
+		log_error("Bad request\n");
+		send_status_response(new_socket, "Invalid redirect response from Fortinet server. VPN could not be established.");
+		return -1;
+	}
+
+	// Extract the id
+	static const char *token_delimiter = " &\r\n";
+	char *next_token = request + strlen(request_head); // strsep does modify the input argument and we don't want to loose our request pointer.
+	char *id_start = strsep(&next_token, token_delimiter);
+
+	if (next_token == NULL) {
+		// In case not found, next_token was set to NULL
+		// This should be invalid because we expect \r\n at the end of the GET request line
+		log_error("Bad request format\n");
+		send_status_response(new_socket, "Invalid formatting of Fortinet server redirect response. VPN could not be established.");
+		return -1;
+	}
+
+	// strsep inserted a NULL at the location of the delimiter.
+	int id_length = strlen(id_start);
+
+	if(id_length == 0 || id_length >= MAX_SAML_SESSION_ID_LENGTH) {
+		log_error("Bad request id\n");
+		send_status_response(new_socket, "Invalid SAML session id received from Fortinet server. VPN could not be established.");
+		return -1;
+	}
+
+	// It was checked above, that the length is smaller than MAX_SAML_SESSION_ID_LENGTH
+	strcpy(id, id_start);
+
+	for (int i = 0; i < id_length; i++) {
+		if (isalnum(id[i]) || id[i] == '-') continue;
+		log_error("Invalid id format\n");
+		send_status_response(new_socket, "Invalid SAML session id received from Fortinet server. VPN could not be established.");
+		return -1;
+	}
+
+	send_status_response(new_socket,
+			"SAML session id received from Fortinet server. VPN will be established...<br>"
+			"You may close this browser tab now.<br>"
+			"<script>"
+			"window.setTimeout(() => { window.close(); }, 5000);\r\n"
+			"document.write(\"<br>This window will close automatically in 5 seconds.\");\r\n"
+			"</script>\r\n");
+	return 0;
+}
+
+/**
+ * Run a http server to listen for SAML login requests
+ *
+ * @return 0 in case of success
+ *         < 0 in case of error
+*/
+int wait_for_http_request(struct vpn_config *config) {
+	int server_fd, new_socket;
+	struct sockaddr_in address;
+	int opt = 1;
+	int addrlen = sizeof(address);
+	long saml_port = config->saml_port;
+
+	// Creating socket file descriptor
+	if ((server_fd = socket(AF_INET, SOCK_STREAM, 0)) == 0) {
+		log_error("Failed to create socket\n");
+		return -1;
+	}
+
+	// Forcefully attaching socket to the port
+	if (setsockopt(server_fd, SOL_SOCKET, SO_REUSEADDR, &opt, sizeof(opt))) {
+		close(server_fd);
+		log_error("Failed to set socket options\n");
+		return -1;
+	}
+
+	address.sin_family = AF_INET;
+	address.sin_addr.s_addr = htonl(INADDR_LOOPBACK);
```

Comment:
It's good to have the socket listen to the loopback interface only, see discussion about this PR.